torch/utils/data/datapipes/map/combinatorics.py

```diff
@@ -44,23 +46,58 @@ def __init__(self,
         super().__init__()
         self.datapipe = datapipe
         self.indices = list(range(len(datapipe))) if indices is None else indices
-        self.index_map = {index_name: num_index for num_index, index_name in enumerate(self.indices)}
-        # We do not lazily shuffle because this way is significantly faster in terms of total time
-        random.shuffle(self.indices)
-
-    def __getitem__(self, index) -> T_co:
-        try:
-            old_numeric_index = self.index_map[index]
-        except KeyError:
-            raise IndexError(f"Index {index} is out of range for {self}.")
-        new_index = self.indices[old_numeric_index]
-        return self.datapipe[new_index]
-
-    # Without __iter__ implemented, by default it tries to use 0-index,
-    # which doesn't work when there is a custom index.
+        self._enabled = True
+        self._seed = None
+        self._rng = random.Random()
+
+    def set_shuffle(self, shuffle=True):
+        self._enabled = shuffle
+        return self
+
+    def set_seed(self, seed: int):
+        self._seed = seed
+        return self
+
     def __iter__(self) -> Iterator[T_co]:
-        for i in self.indices:
-            yield self.datapipe[i]
+        if not self._enabled:
+            for idx in self.indices:
+                yield self.datapipe[idx]
+        else:
+            if self._seed is None:
+                self._seed = int(torch.empty((), dtype=torch.int64).random_().item())
+            self._rng.seed(self._seed)
+            self._seed = None
+            shuffled_indices = self._rng.sample(self.indices, len(self.indices))
+            for idx in shuffled_indices:
+                yield self.datapipe[idx]
```

Comment:
It seems like this is conceptually applying `RandomSampler` to a `MapDataPipe` and thus making it a `IterDataPipe`. That is fine.

I do worry about losing random access to the samples. I suspect users might care about that? Is there anything that users can't do if we convert this to `IterDataPipe`?

If so, we probably want to have `Shuffler` and `RandomSamplerDataPipe` separate.

Reply:
Let's consider a non-0-based index `MapDataPipe`. {"a": 1, "b": 2, "c": 3}
You want to access an element with index "a", do you want `MapDataPipe.shuffle` returns value 1 or any other value? Returning 1 makes shuffling non-functional; returning any other value makes users confused.
And, the order concept for `MapDataPipe` is tied to the order of indices. `__getitem__` function itself should not get involve any order.

Right now, I have two options in mind for `MapDataPipe.shuffle`:
- Like this PR, converting to an `IterDataPipe`
- Keep it as a `MapDataPipe`, but `__getitem__` gives non-shuffled value and `__iter__` gives results based on shuffled indices.

And, only with the above options, can I align the behavior of two types of `DataPipe` to always use different seeds per iteration, which is preferred IMHO.

Reply:
It is not 100% clear to me what users expect when shuffle a custom non-0-index. Do they want the valued shuffled? Or do they just want the ordering returned from `__iter__` to change? I guess we can get rid of the ambiguity can strictly doing one over the other. My take is that users more likely want the ordering to change because the swapping the values among keys isn't that helpful?

> `__getitem__` gives non-shuffled value

This is likely confusing to users, and it is especially problematic if users chain another MapDataPipe operation (e.g. `.batch`) after `.shuffle()` because practically there will be no shuffling done. At that point, making it `IterDataPipe` is better because it should prevent unexpected behaviors.

This is all assuming that users do not want to swap the values among keys when they call `.shuffle()`.